cmd/client/wallet/main.go

```diff
@@ -563,34 +571,46 @@ func convertBalanceIntoReadableFormat(balance *big.Int) string {
 }
 
 // FetchBalance fetches account balance of specified address from the Harmony network
-func FetchBalance(address common.Address) map[uint32]AccountState {
-	result := make(map[uint32]AccountState)
-	for i := 0; i < walletProfile.Shards; i++ {
+func FetchBalance(address common.Address) []*AccountState {
+	result := []*AccountState{}
+	for shardID := 0; shardID < walletProfile.Shards; shardID++ {
+		// Fill in nil pointers for each shard; nil represent failed balance fetch.
+		result = append(result, nil)
+	}
+
+	for shardID := 0; shardID < walletProfile.Shards; shardID++ {
 		balance := big.NewInt(0)
 		var nonce uint64
 
-		result[uint32(i)] = AccountState{balance, 0}
-
-		for retry := 0; retry < rpcRetry; retry++ {
-			server := walletProfile.RPCServer[i][rand.Intn(len(walletProfile.RPCServer[i]))]
-			client, err := clientService.NewClient(server.IP, server.Port)
-			if err != nil {
-				continue
-			}
-
-			log.Debug("FetchBalance", "server", server)
-			response, err := client.GetBalance(address)
-			if err != nil {
-				log.Info("failed to get balance, retrying ...")
-				time.Sleep(200 * time.Millisecond)
-				continue
+		result[uint32(shardID)] = &AccountState{balance, 0}
+
+	LOOP:
+		for j := 0; j < len(walletProfile.RPCServer[shardID]); j++ {
+			for retry := 0; retry < rpcRetry; retry++ {
+				server := walletProfile.RPCServer[shardID][j]
+				client, err := clientService.NewClient(server.IP, server.Port)
+				if err != nil {
+					continue
+				}
+
+				log.Debug("FetchBalance", "server", server)
+				response, err := client.GetBalance(address)
+				if err != nil {
+					log.Info("failed to get balance, retrying ...")
+					time.Sleep(200 * time.Millisecond)
+					continue
+				}
+				log.Debug("FetchBalance", "response", response)
+				respBalance := big.NewInt(0)
+				respBalance.SetBytes(response.Balance)
+				if balance.Cmp(respBalance) < 0 {
```

Comment:
what if some server is slower, and the balance is greater than previous but not the latest? should we loop through the end of the loop and pick the maximum value as latest balance

Reply:
I don't think picking maximum balance is correct. Maximum nonce, might be more correct to reflect more recent transactions.

Reply:
right, maximum nonce

Reply:
Maximum value might not be the correct state as well.

Reply:
yea, we have to rely on consistent nodes to check. This is already working fine. No major problem so far.

Reply:
Ideally the correct way is picking max of (nonce, balance). (In that order) per shard.